iml-gui/crate/src/dependency_tree.rs

```diff
@@ -0,0 +1,444 @@
+        .filter_map(|id| objs.iter().find(|o| o.id() == *id))
+        .map(Arc::clone)
+        .collect()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand_core::{RngCore, SeedableRng};
+    use rand_xoshiro::Xoroshiro64Star;
+    use std::hash::Hash;
+
+    #[derive(Debug, Clone)]
+    struct X {
+        id: u32,
+        deps: Vec<u32>,
+        description: String,
+    }
+
+    impl X {
+        fn new(id: u32, deps: &[u32], desc: &str) -> Self {
+            Self {
+                id,
+                deps: deps.to_vec(),
+                description: desc.to_string(),
+            }
+        }
+    }
+
+    impl Deps<u32> for X {
+        fn id(&self) -> u32 {
+            self.id
+        }
+        fn deps(&self) -> &[u32] {
+            &self.deps
+        }
+        fn has(&self, k: &u32) -> bool {
+            self.deps.contains(k)
+        }
+    }
+
+    // Note: Y cannot be Deps by itself, because Deps::deps() returns slice
+    #[derive(Debug, Clone)]
+    struct Y {
+        id: u32,
+        deps: Vec<String>,
+        description: String,
+    }
+
+    impl Y {
+        fn new(id: u32, str_deps: &[&str], desc: &str) -> Self {
+            Self {
+                id,
+                deps: str_deps.iter().map(|s| s.to_string()).collect(),
+                description: desc.to_string(),
+            }
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    struct Context<K: Hash + Eq + Debug> {
+        visited: HashSet<K>,
+        indent: usize,
+        is_new: bool,
+    }
+
+    fn shuffle<R: RngCore, T>(slice: &mut [T], rng: &mut R) {
+        for i in (1..slice.len()).rev() {
+            let j = (rng.next_u64() as usize) % (i + 1);
+            slice.swap(i, j);
+        }
+    }
+
+    #[test]
+    fn build_dependency_tree_test() {
+        let x_list: Vec<X> = vec![
+            X::new(39, &[], "Install packages on server oss2.local."),
+            X::new(40, &[39], "Configure NTP on oss2.local."),
+            X::new(41, &[39], "Enable LNet on oss2.local."),
+            X::new(42, &[39], "Configure Corosync on oss2.local."),
+            X::new(43, &[42], "Start Corosync on oss2.local"),
+            X::new(44, &[41], "Load the LNet kernel modules."),
+            X::new(45, &[44], "Start the LNet networking layer."),
+            X::new(46, &[39, 43], "Configure Pacemaker on oss2.local."),
+            X::new(47, &[43, 46], "Start Pacemaker on oss2.local."),
+            X::new(48, &[39, 40, 41, 42, 45, 46, 47], "Setup managed host oss2.local."),
+        ];
+        let x_arcs: Vec<Arc<X>> = x_list.into_iter().map(Arc::new).collect();
+
+        let dag = build_direct_dag(&x_arcs);
+        let result = build_dag_str(&dag, &x_to_string).join("\n");
+        assert_eq!(result, TREE_DIRECT);
+
+        let dag = build_inverse_dag(&x_arcs);
+        let result = build_dag_str(&dag, &x_to_string).join("\n");
+        assert_eq!(result, TREE_INVERSE);
+    }
+
+    #[test]
+    fn cyclic_dependency() {
```

Comment:
If DAG stands for "Directed acyclic graph", why do you test cyclic deps? :)

Reply:
It is true, although the absence of cycles is not enforced by types or something, and we try to build trees for the external data.